sklearn/base.py

```diff
@@ -544,10 +566,80 @@ def fit_predict(self, X, y=None):
         return self.fit(X).predict(X)
 
 
+class OneToOneMixin(object):
+    """Provides get_feature_names for simple transformers
+
+    Assumes there's a 1-to-1 correspondence between input features
+    and output features.
+    """
+
+    def get_feature_names(self, input_features=None):
+        """Get feature names for transformation.
+
+        Returns input_features as this transformation
+        doesn't add or drop features.
+
+        Parameters
+        ----------
+        input_features : array-like of string
+            Input feature names.
+
+        Returns
+        -------
+        feature_names : array-like of string
+            Transformed feature names
+        """
+        if input_features is not None:
+            return input_features
+        else:
+            raise ValueError("Don't know how to get"
+                             " input feature names for {}".format(self))
+
+
+def _get_sub_estimators(est):
```

Comment:
Shouldn't we solve this issue by adding a "sub_estimators_" attribute on all meta-estimators?

Reply:
Why is that easier / better?

Reply:
Well, it's the meta-estimator's responsibility to define what it's sub-estimators are.

The list above should be modified when we add meta-estimators. It cannot be extended by other packages implementing meta-estimators.

I find that in an object-oriented design, it is more natural that each class deals with specifying it's own functionality.

Reply:
if we add meta-estimators that don't obey our conventions a test will fail.
Other packages implementing meta-estimators should either obey our conventions or they can implement their own ``get_feature_names``.
If you wanted me to implement this on the meta-estimator, I would probably just make this a method of the MetaEstimatorMixin. Then other packages / estimators could overwrite that method instead of overwriting ``get_feature_names``. I don't think that's better.

Reply:
Indeed, there are several ways to delegate the responsibility to the meta-estimator, either via delegating to a method, or to an attribute. The attribute seemed pretty light to me.

With regards to avoiding premature test failures, we can simple pass when sub_estimator_ is not present.

Reply:
a new sub-estimator was added while this discussion was happening lol.
The reason why I don't like the attribute is because it makes our API contract more verbose. I'm ok with doing it, but in the end this will make it harder for people to contribute by increasing the number of required attributes (I'd rather add n_input_features_ and n_output_features_ first ;).

I'm not sure I understand your last sentence.

I don't really care that much in either direction here, it's an implementation detail that's not really important to the larger feature addition, I think.